File: samples/ppocr/text_cls.cpp

/*
 * Copyright (C) 2026 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/text_cls.hpp"

#include <sstream>

#include "common/cmdline.hpp"
#include "common/model_profiler.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("model_prefix", 'm',
                        "model prefix of the model suite files", false,
                        "/opt/vastai/vaststreamx/data/models/"
                        "textline_ori_fp16_1-3-80-160/mod");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "../data/configs/textline_ori_rgbplanar.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input image file", false,
                        "../data/images/word_336.png");
  args.add<std::string>("dataset_val_file", '\0', "dataset validation file",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  const int batch_size = 1;

  vsx::TextClassifier model(args.get<std::string>("model_prefix"),
                            args.get<std::string>("vdsp_params"), batch_size,
                            args.get<uint32_t>("device_id"),
                            args.get<std::string>("hw_config"));

  auto image_format = model.GetFusionOpIimageFormat();

  if (args.get<std::string>("dataset_val_file").empty()) {
    vsx::Image vsx_image;
    vsx::MakeVsxImage(args.get<std::string>("input_file"), vsx_image,
                      image_format);
    auto tensor = model.Process(vsx_image);
    if (tensor.Data<float>()[0] > tensor.Data<float>()[1]) {
      std::cout << "Text Line angle: 0, confidence: " << tensor.Data<float>()[0]
                << std::endl;
    } else {
      std::cout << "Text Line angle: 180, confidence: "
                << tensor.Data<float>()[1] << std::endl;
    }

  } else {
    std::vector<std::string> val_lines =
        vsx::ReadFileList(args.get<std::string>("dataset_val_file"));

    std::vector<std::string> filelist;
    std::vector<int> gt_indies;
    for (auto& line : val_lines) {
      std::istringstream iss(line);
      std::string file;
      int index;
      if (iss >> file >> index) {
        filelist.push_back(file);
        gt_indies.push_back(index);
      } else {
        std::cerr << "Parsing dataset val file Failed. line:" << line
                  << std::endl;
        return -1;
      }
    }
    auto dataset_root = args.get<std::string>("dataset_root");
    int correct_count = 0;
    for (size_t i = 0; i < filelist.size(); i++) {
      auto fullname = filelist[i];
      if (!dataset_root.empty()) fullname = dataset_root + "/" + filelist[i];
      std::cout << fullname << std::endl;
      vsx::Image vsx_image;
      vsx::MakeVsxImage(fullname, vsx_image, image_format);
      auto tensor = model.Process(vsx_image);
      int index = 0;
      if (tensor.Data<float>()[0] < tensor.Data<float>()[1]) index = 1;

      if (index == gt_indies[i]) {
        correct_count++;
      }
      std::cout << "Predicted index: " << index
                << ", GT index: " << gt_indies[i]
                << ", correct_count: " << correct_count << std::endl;
    }

    std::cout << "Accuracy: " << correct_count / float(filelist.size())
              << std::endl;
  }

  return 0;
}

File: samples/ppocr/text_cls.py
#
# Copyright (C) 2026 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

from common.text_cls import TextClassifier
import common.utils as utils

import cv2
import argparse


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m",
        "--model_prefix",
        default="/opt/vastai/vaststreamx/data/models/textline_ori_fp16_1-3-80-160/mod",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--vdsp_params",
        default="./data/configs/textline_ori_rgbplanar.json",
        help="vdsp preprocess parameter file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/word_336.jpg",
        help="input file",
    )
    parser.add_argument(
        "--dataset_val_file",
        default="",
        help="dataset validation file",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="input dataset root",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    label_list = ["0", "180"]
    batch_size = 1

    model = TextClassifier(args.model_prefix,  args.vdsp_params, label_list, batch_size, args.device_id)
    image_format = model.get_fusion_op_iimage_format()

    if args.dataset_val_file == "":
        cv_image = cv2.imread(args.input_file)
        assert cv_image is not None, f"Failed to read input file: {args.input_file}"
        vsx_image = utils.cv_bgr888_to_vsximage(cv_image, image_format, args.device_id)
        angle, score = model.process(vsx_image)
        print(
            f"Image angle: {angle}, confidence: {score:.4f}"
        )
    else:
        filelist = []
        gt={}
        with open(args.dataset_val_file, "rt") as f:
            lines = f.readlines()
            for line in lines:
                filename, label = line.strip().split()
                filelist.append(filename)
                gt[filename] = int(label)        
        correct_count = 0
    
        for file in filelist:
            fullname = os.path.join(args.dataset_root, file)
            print(fullname)
            cv_image = cv2.imread(fullname)
            assert cv_image is not None, f"Failed to read input file: {fullname}"
            vsx_image = utils.cv_bgr888_to_vsximage(
                cv_image, image_format, args.device_id
            )
            angle, score = model.process(vsx_image)
            predicted_label = label_list.index(str(angle))
            if predicted_label == gt.get(file, -1):
                correct_count += 1
            print(f"Image: {file}, Label: {gt.get(file, -1)}, Predicted: {predicted_label}, Correct: {correct_count}")
        print(f"Accuracy: {correct_count / len(filelist):.4f}")